### ISSUE TITLE
1.3.40-eap-67: Assertion failed in ObjCExportHeaderGenerator

### ISSUE THREAD

**Issue by jonasbark** [JetBrains/kotlin-native#3053]
```
10:51:07.794 [ERROR] [system.err] e: Compilation failed: Assertion failed
10:51:07.794 [ERROR] [system.err] 
10:51:07.795 [ERROR] [system.err]  * Source files: ApiException.kt, CoroutineUtils.kt, ConnectApi.kt, ApiAuthorization.kt, UserAgent.kt, ModelValidationExtensions.kt, RedactedApi.kt, Validation.kt, ValidationBuilder.kt, ValidationResult.kt, Builder.kt, Validation.kt, JsonSchema.kt, CoroutineUtilsImpl.kt, SCExtensions.kt, Ios.kt, actualHttpClient.kt
10:51:07.795 [ERROR] [system.err]  * Compiler version info: Konan: 1.3-eap-10415 / Kotlin: 1.3.40
10:51:07.795 [ERROR] [system.err]  * Output kind: FRAMEWORK
10:51:07.795 [ERROR] [system.err] 
10:51:07.795 [ERROR] [system.err] e: java.lang.AssertionError: Assertion failed
10:51:07.795 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.referenceClass(ObjCExportHeaderGenerator.kt:81)
10:51:07.795 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.mapReferenceTypeIgnoringNullability$backend_native_compiler(ObjCExportHeaderGenerator.kt:671)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.mapReferenceType$backend_native_compiler(ObjCExportHeaderGenerator.kt:599)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.mapType(ObjCExportHeaderGenerator.kt:737)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.access$mapType(ObjCExportHeaderGenerator.kt:44)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl$buildMethod$1.invoke(ObjCExportHeaderGenerator.kt:500)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.buildMethod(ObjCExportHeaderGenerator.kt:521)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportTranslatorImpl.translateClass(ObjCExportHeaderGenerator.kt:189)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportHeaderGenerator.generateClass$backend_native_compiler(ObjCExportHeaderGenerator.kt:1033)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportHeaderGenerator$translateModule$4.invoke(ObjCExportHeaderGenerator.kt:929)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExportHeaderGenerator.translateModule(ObjCExportHeaderGenerator.kt:941)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExport.produceInterface(ObjCExport.kt:67)
10:51:07.796 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.objcexport.ObjCExport.<init>(ObjCExport.kt:44)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.ToplevelPhasesKt$objCExportPhase$1.invoke(ToplevelPhases.kt:69)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.ToplevelPhasesKt$objCExportPhase$1.invoke(ToplevelPhases.kt)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$namedOpUnitPhase$1.invoke(PhaseBuilders.kt:116)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$namedOpUnitPhase$1.invoke(PhaseBuilders.kt:114)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
10:51:07.797 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:41)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:27)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:79)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:34)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:84)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:42)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:104)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:82)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:50)
10:51:07.798 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:215)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:207)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:215)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:212)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:40)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.konan.util.UtilKt.profile(Util.kt:34)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:214)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:296)
10:51:07.799 [ERROR] [system.err]       at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:16)
10:51:07.799 [ERROR] [system.err] 
10:51:07.878 [DEBUG] [org.gradle.process.internal.DefaultExecHandle] Changing state to: FAILED
10:51:07.878 [DEBUG] [org.gradle.process.internal.DefaultExecHandle] Process 'command '/Library/Java/JavaVirtualMachines/jdk1.8.0_201.jdk/Contents/Home/bin/java'' finished with exit value 2 (state: FAILED)
10:51:07.879 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationExecutor] Completing Build operation 'Execute compile for :Redacted:linkDebugFrameworkIosX64'
10:51:07.879 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationExecutor] Build operation 'Execute compile for :Redacted:linkDebugFrameworkIosX64' completed
10:51:07.879 [DEBUG] [org.gradle.api.internal.tasks.execution.DefaultTaskFingerprinter] Fingerprinting property destinationDir (Output) for task ':Redacted:linkDebugFrameworkIosX64'
10:51:07.880 [DEBUG] [org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter] Removed task artifact state for {} from context.
10:51:07.880 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationExecutor] Completing Build operation 'Task :Redacted:linkDebugFrameworkIosX64'
10:51:07.869 [LIFECYCLE] [class org.gradle.internal.buildevents.TaskExecutionLogger] 
10:51:07.869 [LIFECYCLE] [class org.gradle.internal.buildevents.TaskExecutionLogger] > Task :Redacted:linkDebugFrameworkIosX64 FAILED
```

Using: 
```
kotlin_version=1.3.40-eap-67
coroutines_version=1.2.1-1.3.40-eap-67
ktor_version=1.2.1-1.3.40-eap-67
serialization_version=0.11.1-1.3.40-eap-67-2
```
Please let me know if additional information is needed and how I can provide those.

**Comment by SvyatoslavScherbina** [comment]
Does this happen when you compile with 1.3.31?
Looks much like duplicate of #2120.

**Comment by jonasbark** [comment]
With 1.3.31 I'm only getting as far as https://github.com/JetBrains/kotlin-native/issues/3019 so I can't confirm if it's a duplicate or not.

**Comment by SvyatoslavScherbina** [comment]
You probably have some public method or class with generics having non-public upper bound, like in [KT-27112](https://youtrack.jetbrains.com/issue/KT-27112).
This class or method causes the compiler error.

**Comment by jonasbark** [comment]
I did not find anything like that.
However I could track it down to where I defined our serializable models - here's a full example:

```kotlin
/**
 * Represents the native date. On JVM it's java.util.Date, on iOS it's NSDate.
 * @see NativeDate
 */
typealias Date = NativeDate

private fun String.verifyParam(s: String, function: (String) -> Boolean) {
    if (!function(this)) {
        throw IllegalArgumentException("$this is not accepted for $s")
    }
}

private val json = Json(JsonConfiguration.Stable.copy(encodeDefaults = false))

@Serializable
data class TestModel(
    val name: String? = null,
    @Serializable(with = NativeDateSerializer::class) val created: Date? = null,
    val id: String? = null,
    val version: String? = null
) {
    init {
        classifier?.verifyParam("classifier") { it in setOf("...") }
    }

    /**
     * serializes this object to a JSON string
     * @return JSON representation of this object
     */
    fun toJson() = json.stringify(serializer(), this)

    companion object {
        /**
         * deserializes the given JSON to an object of this kind
         * @param string JSON string
         * @return deserialized object
         */
        fun fromJson(string: String) = json.parse(serializer(), string)
    }
}
```
No generics here so it might really be a kotlinx-serialization issue.

**Comment by SvyatoslavScherbina** [comment]
Thanks for the tracking, this helps a lot!
I'm able to reproduce this.

**Comment by SvyatoslavScherbina** [comment]
Will be fixed in 1.3.40 EAP 3.

**Comment by jonasbark** [comment]
It works, thank you!
